libzfs_core/test/test_nvlist.py

```diff
@@ -0,0 +1,637 @@
+        self._assertIntDictsEqual(props, res)
+
+    def test_explicit_int32_too_large_value(self):
+        with self.assertRaises(OverflowError):
+            props = {"key": int32_t(2 ** 31)}
+            res = self._dict_to_nvlist_to_dict(props)
+
+    def test_explicit_int32_too_small_value(self):
+        with self.assertRaises(OverflowError):
+            props = {"key": int32_t(-(2 ** 31) - 1)}
+            res = self._dict_to_nvlist_to_dict(props)
+
+    def test_explicit_int16_value(self):
+        props = {"key": int16_t(1)}
+        res = self._dict_to_nvlist_to_dict(props)
+        self._assertIntDictsEqual(props, res)
+
+    def test_explicit_int16_max_value(self):
+        props = {"key": int16_t(2 ** 15 - 1)}
+        res = self._dict_to_nvlist_to_dict(props)
+        self._assertIntDictsEqual(props, res)
+
+    def test_explicit_int16_min_value(self):
+        props = {"key": int16_t(-(2 ** 15))}
+        res = self._dict_to_nvlist_to_dict(props)
+        self._assertIntDictsEqual(props, res)
+
+    def test_explicit_int16_too_large_value(self):
+        with self.assertRaises(OverflowError):
+            props = {"key": int16_t(2 ** 15)}
+            res = self._dict_to_nvlist_to_dict(props)
+
+    def test_explicit_int16_too_small_value(self):
+        with self.assertRaises(OverflowError):
+            props = {"key": int16_t(-(2 ** 15) - 1)}
+            res = self._dict_to_nvlist_to_dict(props)
+
+    def test_explicit_int8_value(self):
+        props = {"key": int8_t(1)}
+        res = self._dict_to_nvlist_to_dict(props)
+        self._assertIntDictsEqual(props, res)
+
+    def test_explicit_int8_max_value(self):
+        props = {"key": int8_t(2 ** 7 - 1)}
+        res = self._dict_to_nvlist_to_dict(props)
+        self._assertIntDictsEqual(props, res)
+
+    def test_explicit_int8_min_value(self):
+        props = {"key": int8_t(-(2 ** 7))}
+        res = self._dict_to_nvlist_to_dict(props)
+        self._assertIntDictsEqual(props, res)
+
+    def test_explicit_int8_too_large_value(self):
+        with self.assertRaises(OverflowError):
+            props = {"key": int8_t(2 ** 7)}
+            res = self._dict_to_nvlist_to_dict(props)
+
+    def test_explicit_int8_too_small_value(self):
+        with self.assertRaises(OverflowError):
+            props = {"key": int8_t(-(2 ** 7) - 1)}
+            res = self._dict_to_nvlist_to_dict(props)
+
+    def test_nested_dict(self):
+        props = {"key": {}}
+        res = self._dict_to_nvlist_to_dict(props)
+        self.assertEqual(props, res)
+
+    def test_nested_nested_dict(self):
+        props = {"key": {"key": {}}}
+        res = self._dict_to_nvlist_to_dict(props)
+        self.assertEqual(props, res)
+
+    # XXX API deficiency: C code allows empty arrays,
+    # but it is impossible to deduce intended value type
+    # from an empty array and there is no way to provide
+    # it explicitly.
+    @unittest.expectedFailure
```

Comment:
I think the "expected failure" feature of unittest is a misfeature.  What you really want are issues filed in an issue tracker for each of these desirable-but-unacheived behaviors.  The issues might even have the code for each of these tests.  But there's not really any use having them checked in and run (and their results ignored, at best) with the test suite.